### PR TITLE
Make it possible to call functions passed as the Slot's fillProps in the Fill

### PR DESCRIPTION
**Pull request by IreneStr**
## Description
This functionality is needed for the more menu implementation in https://github.com/WordPress/gutenberg/pull/5729. We need a way to pass props to from Slots to Fills. For example, we'd like to pass the `onClose` callback function that is defined in the more menu Slot to be passed to the more menu Fill. 

#### Slot
```jsx
<Slot name="moreMenu" fillProps={ { onClose: someFunction } } />
```

#### Fill
```jsx
<Fill name="moreMenu">
	{ ( props ) => {
		return <Button onClick={ props.onClose() }/>
	} };
</Fill>
```

## How Has This Been Tested?

This has been tested with `bubblesVirtually` in the browser, but no unit-tests were written because `enzyme` doesn't support portals. We pasted the following snippet in `edit-post/components/layout/index.js` for that:
```
<Slot name="testSlot" fillProps={ { onClose: () => { console.log( "hoi" ) } } }/>
<Fill name="testSlot">
    { ( props ) => { return <button onClick={ props.onClose }>Click me</button>; } }
</Fill>
```

Unit-tests were written for this enhancement without `bubblesVirtually`.

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x]  My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by aduth** [comment]
Context? Reasoning?

**Comment by IreneStr** [comment]
@aduth I've added a description. Please let me know if you need more info. The PR itself is still very much WIP, btw.

**Comment by aduth** [comment]
I've been spending some time looking at this, and your note about testing `bubblesVirtually` had me interested, because from what I can tell, Enzyme will happily test portaled content. Turns out, there was a bug where the render order of the `<Slot bubblesVirtually />` had to occur after the Fill is rendered, which probably explains the errors you were seeing. 83c9921 introduces identical testing between `bubblesVirtually: true` and `bubblesVirtually: false`, including a fix to a failing test where virtual behavior was not rendering in the correct order (see extended message of 83c9921). 50604e7 then resolves the issue where we had depended upon a particular render order of Slot vs. Fill.

I am still going to spend some time looking at slots behavior, notably the issue described in #5633, which unfortunately is still present even after these changes, but also exists on master. For the purposes of these changes, it may be fine to proceed as-is.

**Comment by aduth** [comment]
Actually, I'm noticing an issue related to the change to use fragments. It appears since we pass an array of child fills, we must ensure a valid key is assigned, else a warning occurs (noticeable when reproducing steps from #5633). This existed previously (in the non-virtual implementation only), but was dropped thinking Fragment alone was sufficient.

**Comment by aduth** [comment]
Minimal demonstration of the issue: https://codepen.io/aduth/pen/dmRpYj

**Comment by aduth** [comment]
I'm still iterating on some improvements here to the most recently introduced issues (the demonstration is actually inaccurate, and while it's still an issue, it's actually a deeper problem relating to the occurrence order of fills when a fill is unmounted). I plan to push some changes this evening. I've had success in reproducing via unit tests and resolving two additional bugs with Slot/Fill, including #5633 and the aforementioned issue. I'm not sure yet what of this I'll include in this branch, or if a separate pull request is necessary.

**Review by aduth — APPROVED** [review]
My local changes are mostly in working shape, but it's grown a bit unwieldy to consider introducing here. I'll plan instead to open a separate pull request from this one, and have reverted my last two commits. This otherwise looks fine to me (notwithstanding the now-known bugs of Slot/Fill which exist irrespective these changes).

File: components/slot-fill/slot.js [hunk 1]
```diff
@@ -45,7 +45,7 @@ class Slot extends Component {
 	}
 
 	render() {
-		const { name, bubblesVirtually = false } = this.props;
+		const { name, bubblesVirtually = false, fillProps } = this.props;
```
Comment (gziolo) [inline components/slot-fill/slot.js]: Should `fillProps` default to `{}` when not provided.

File: components/slot-fill/test/slot.js [hunk 2]
```diff
@@ -82,6 +82,27 @@ describe( 'Slot', () => {
 		expect( element.find( 'Slot > div' ).html() ).toBe( '<div><span></span><div></div>text</div>' );
 	} );
 
+	it( 'calls the functions passed as the Slot\'s fillProps in the Fill', () => {
+		const onClose = jest.fn();
+
+		const element = mount(
+			<Provider>
+				<Slot name="chicken" fillProps={ { onClose: onClose } } />
```
Comment (gziolo) [inline components/slot-fill/test/slot.js]: `{ onClose: onClose }` can be simplified o `{ onClose }`

File: components/slot-fill/test/slot.js [hunk 3]
```diff
@@ -82,6 +82,27 @@ describe( 'Slot', () => {
...
+							<button onClick={ props.onClose }>Click me</button>
+						);
+					} }
+				</Fill>
+			</Provider>
+		);
+
+		element.find( 'button' ).simulate( 'click' );
+
+		expect( onClose ).toHaveBeenCalled();
```
Comment (gziolo) [inline components/slot-fill/test/slot.js]: `expect( onClose ).toHaveBeenCalledTimes( 1 )` - might be a better fit here.

**Review by gziolo — APPROVED** [review]
I had some minor comments, otherwise it looks good to go 👍

**Pull request merged**